### PR TITLE
Add archive-include config

### PR DESCRIPTION
**Pull request by rukai**
Progress towards https://github.com/nextest-rs/nextest/issues/1340

This PR implements the design given by rain in https://github.com/nextest-rs/nextest/issues/1340#issuecomment-1972163611
However it only implements the smallest possible functionality that will the allow the full design to be implemented in a backwards compatible manner in the future.
This is because I would like to land a small part of the design to ensure I am on track and also because I cant afford the time to implement all of these features right now.

The full implementation will look like:
```toml
[profile.ci]
archive-include = [
    { path = "foo", relative-to = "target" },
    { path = "bar/*", relative-to = "target-profile" },
]
```

However this PR's config will only support:
```toml
[profile.ci]
archive-include = [
    { path = "foo", relative-to = "target" },
    { path = "bar", relative-to = "target" },
]
```
This leaves the following features for follow up work:
* `relative-to = "target-profile"`
    + This can easily be introduced later.
    + For now, the user can also set `{ path = "debug/foo", relative-to = "target" }, { path = "release/foo", relative-to = "target" }` for a hacky work around.
* `*` globbing
    + while adding it later would technically be a breaking change since `*` can be included in a valid file name on linux. Using `*` in file names is generally considered a very bad idea, so I'm not worried about breakage here. I am however happy to include globbing support in this PR if such breakage is unacceptable.

## `optional` config

It was also stated that by default the `nextest archive` command should fail when the path to include does not exist.
I considered the positives and negatives of this versus just ignoring missing paths:
* Positives
    + Will give slightly better diagnostics to what went wrong if they enter the wrong path
       + I dont believe this will be a huge improvement, since the user has already diagnosed that there is something missing from their archived target that they need to configure nextest to pick up. So if after configuring nextest to include the path they get the same issue due to missing artifacts, they will naturally inspect the archive and config to find out what is going wrong.
    + Will catch paths that no longer need to be included and can be removed from the config.
* Negatives
   + Will cause archiving to unexpectedly fail in these scenarios:
        - nextest config is workspace wide and only a single crate produces the build artifact
        - When a crate or its dependency is designed to only output the build artifact in only some cases.

Ultimately the negatives are not huge issues since adding `optional = true` is not that hard, its mostly just that it feels `optional = true` is the more reasonable default at which point having it configurable seems less valuable.
In conclusion, I dont know which is better, so I've taken the simpler option of not implementing the `optional` config, but just confirm that you do want fail by default and I'll go ahead and implement it in this PR.

## Questions

* Am I expected to implement custom diagnostics for the ArchiveInclude deserialization
* should append_path_recursive be given an optional recursion depth so it can be used instead of append_dir_one_level?
* Is there any extra documentation I need to add? Or is the automatically generated page at https://nexte.st/book/configuration.html sufficient?

### CODE REVIEW

**Comment by sunshowers** [comment]
Thanks! Is this ready for review? If so I'll try and make time in a few days (really busy with other stuff this week)

**Comment by sunshowers** [comment]
Planning to look at this tomorrow!

File: integration-tests/tests/integration/main.rs [hunk 1]
```diff
@@ -338,13 +338,66 @@ fn test_relocated_run() {
...
+    let (_p2, extracted_target) = run_archive(&archive_file);
+
+    let expected_path = extracted_target
+        .join("application-data")
+        .join("foo")
+        .join("bar")
+        .join("some_file.txt");
+
+    assert!(
+        !Path::new(&expected_path).join("application-data").exists(),
```
Comment (sunshowers) [inline integration-tests/tests/integration/main.rs:354]: Can be `Utf8Path`.

File: nextest-runner/src/config/config_impl.rs [hunk 2]
```diff
@@ -702,6 +702,13 @@ impl<'cfg> NextestProfile<'cfg, FinalConfig> {
             .unwrap_or(self.default_profile.fail_fast)
     }
 
+    /// Returns the archive-include config for this profile.
+    pub fn archive_include(&self) -> &Vec<ArchiveInclude> {
```
Comment (sunshowers) [inline nextest-runner/src/config/config_impl.rs:706]: `&[ArchiveInclude]`

File: nextest-runner/src/reuse_build/archiver.rs [hunk 3]
```diff
@@ -114,16 +117,19 @@ struct Archiver<'a, W: Write> {
...
 impl<'a, W: Write> Archiver<'a, W> {
+    #[allow(clippy::too_many_arguments)]
     fn new(
         binary_list: &'a BinaryList,
         cargo_metadata: &'a str,
         path_mapper: &'a PathMapper,
         format: ArchiveFormat,
         compression_level: i32,
         writer: W,
+        profile: &NextestProfile<'_, FinalConfig>,
```
Comment (sunshowers) [inline nextest-runner/src/reuse_build/archiver.rs:132]: `writer` should be the last argument as it's an out param. I think `profile` should maybe be the first argument.

File: nextest-runner/src/reuse_build/archiver.rs [hunk 4]
```diff
@@ -295,6 +316,46 @@ impl<'a, W: Write> Archiver<'a, W> {
...
 
+    fn append_path_recursive(
+        &mut self,
+        rel_path: &Utf8Path,
+        src_path: &Utf8Path,
+    ) -> Result<(), ArchiveCreateError> {
+        let entries =
+            src_path
+                .read_dir_utf8()
+                .map_err(|error| ArchiveCreateError::InputFileRead {
```
Comment (sunshowers) [inline nextest-runner/src/reuse_build/archiver.rs:327]: We should indicate here that the file read is user-specified and not something that nextest archives by default.

**Review by sunshowers — COMMENTED** [review]
This looks awesome! Just a few minor comments.

I'll make these changes myself and land this. Thanks again!

**Comment by sunshowers** [comment]
Paths being optional by default still concerns me. I think I'll change that to required in a followup, so that.

Regarding docs, we should add a section to https://nexte.st/book/reusing-builds.html. I'll try writing some docs up tomorrow, unless you want to give it a shot.

I'm hoping to release this within the next 1-2 days. I tend to release mornings or afternoons Pacific time because it's better for breakages to be discovered quickly.

**Comment by sunshowers** [comment]
Regarding recursion depth -- yeah, I think `append_path_recursive` should have an `Option<usize>` depth param.

File: nextest-runner/src/reuse_build/archiver.rs (hunk 4 above)
Comment (sunshowers) [reply nextest-runner/src/reuse_build/archiver.rs:327]: Will do this in a followup.

File: nextest-runner/src/config/archive_include.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
+// Copyright (c) The nextest Contributors
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use serde::Deserialize;
+
+/// Type for the archive-include key.
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+pub struct ArchiveInclude {
+    pub(crate) path: String,
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply nextest-runner/src/config/archive_include.rs:10]: 1 done.

**Comment by sunshowers** [comment]
Having thought about it more, I think I'd rather just produce a warning if a path wasn't found. We're going to produce warnings for a few other things too.

**Comment by sunshowers** [comment]
Writing down some TODOs here based on what I saw:

- [x] Add optional depth param, and limit depth by default to a reasonable value (we've chosen 8): https://github.com/nextest-rs/nextest/pull/1411
- [ ] Warn if a path isn't present (may allow this to be configured as required or skipped in the future) -- also include warn if a depth = 0 is a directory, which is similar
- [ ] Check that path is well-formed
- [ ] Add test for skipped file types (not file, dir or symlink)
- [ ] For errors, indicate why we're deciding to include a file in the archive (using an enum)
- [ ] Write docs

Getting these all sorted will take me a few hours, so the release will probably be later this week. If you want to take on some of them then I'd love that -- will aim to get any followups in quickly and get this out soon.

**Comment by rukai** [comment]
@sunshowers thanks for getting this over the line!
I might be able to find time to take on one or two of those items on friday but if you get to them first thats fine by me.

**Comment by sunshowers** [comment]
I fell pretty sick the last two days so won't be able to finish it up this week (need to catch up on work) -- feel free to tackle as many of the above as you want.
